src/psyclone/psyir/symbols/symbol_table.py

```diff
@@ -546,6 +546,182 @@ def add(self, new_symbol, tag=None):
+                            self.next_available_name(
+                                csym.name, other_table=other_table))
+                    # We can then add an import from the Container.
+                    self.add(csym)
+                else:
+                    # If there is a wildcard import from this container then
+                    # we'll need that in this Table too.
+                    if csym.wildcard_import:
+                        self_csym.wildcard_import = True
+            else:
+                self.add(csym)
+            # We must update all references to this ContainerSymbol
+            # so that they point to the one in this table instead.
+            imported_syms = other_table.symbols_imported_from(csym)
+            for isym in imported_syms:
+                if isym.name in self:
+                    # We have a potential clash with a symbol imported
+                    # into the other table.
+                    callsite_sym = self.lookup(isym.name)
+                    if not callsite_sym.is_import:
+                        # The calling merge() method has already checked that
+                        # we don't have a clash between symbols of the same
+                        # name imported from different containers. We don't
+                        # support renaming an imported symbol but the
+                        # symbol in this table can be renamed so we do that.
+                        self.rename_symbol(
+                            callsite_sym,
+                            self.next_available_name(
+                                callsite_sym.name, other_table=other_table))
+                isym.interface = ImportInterface(self.lookup(csym.name))
+
+    def _add_symbols_from_table(self, other_table, include_arguments=True):
+        '''
+        Takes symbols from the supplied symbol table and adds them to this
+        table. _add_container_symbols_from_table() must have been called
+        before this method in order to handle any Container Symbols and update
+        those Symbols imported from them.
+
+        :param other_table: the symbol table from which to add symbols.
+        :type other_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param bool include_arguments: whether or not to include symbols that \
+                                       are formal routine arguments.
+
+        :raises InternalError: if an imported symbol is found that has not \
+            been updated to refer to a Container in this table.
+
+        '''
+        if include_arguments:
+            symbols_to_skip = []
+        else:
+            symbols_to_skip = other_table.argument_list[:]
+
+        try:
+            # In the case where the 'other_table' belongs to a routine,
+            # we don't want or need the symbol representing that routine.
+            rsym = other_table.lookup_with_tag("own_routine_symbol")
+            if isinstance(rsym, RoutineSymbol):
```

Comment:
Should we check that it already exist and is a RoutineSymbol in self?
(only if this doesn't cause many new errors, we may have inconsistent trees in the tests that a Routine nodes does not also imply a declaration)

Reply:
I tried that and it caused a lot of problems in the Fortran backend (I think because there we create a brand-new Container and then add Symbols to it) so I reverted my changes.